File: 프로그래머스/jsh/js/오픈채팅방.js
/**
 * 프로그래머스 - 오픈채팅방 솔루션
 * https://programmers.co.kr/learn/courses/30/lessons/42888?language=javascript
 * 사용자 정보(uid, nickname)를 관리하는 Map 자료구조 사용
 * command에 따른 map update
 * 업데이트한 내용에 따라 command와 uid를 저장하고 탐색하면서 로그 저장
 * @param {*} record ["command, uid, nickname"] 형태의 문자열을 저장하는 배열 
 * @returns 최종 접속 로그를 반환
 */
function solution(record) {
  let answer = [];
  const userDB = new Map();
  const events = [];
  record.forEach(message => {
      const keywordList = message.split(' ');
      const [command, uid] = keywordList;
      if (command === "Enter" || command === "Change") {
          const nickname = keywordList[2];
          userDB.set(uid, nickname);
      }
      events.push([command, uid]);
      
  })
  events.forEach(event => {
      const [command, uid] = event;
      if (command === "Enter") {
          answer.push(`${userDB.get(uid)}님이 들어왔습니다.`);
      } else if (command === "Leave") {
          answer.push(`${userDB.get(uid)}님이 나갔습니다.`);
      }
  })
  return answer;
}

File: 프로그래머스/jsh/python/오픈채팅방.py
def solution(record):
    answer = []
    # DB개념으로 딕셔너리 자료형 사용 (Map)
    userDB = dict()
    events = []
    for message in record:
        keywordList = message.split(' ');
        command, uid = keywordList[0], keywordList[1]
        # Enter Change command 상관없이 DB를 업데이트
        if command == "Enter" or command == "Change":
            nickname = keywordList[2]
            userDB[uid] = nickname
        events.append((command, uid))
    
    for event in events:
        command, uid = event[0], event[1]
        if command == "Enter":
            answer.append(f'{userDB[uid]}님이 들어왔습니다.')
        elif command == "Leave":
            answer.append(f'{userDB[uid]}님이 나갔습니다.')
    
    return answer